File: app/controllers/facility/RoomController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.facility;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import controllers.base.BaseController;
import controllers.iop.transfer.api.ExternalFacilityAPI;
import io.ebean.DB;
import io.ebean.ExpressionList;
import io.ebean.text.PathProperties;
import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.stream.StreamSupport;
import javax.inject.Inject;
import miscellaneous.config.ConfigReader;
import miscellaneous.datetime.DateTimeHandler;
import models.calendar.DefaultWorkingHours;
import models.calendar.ExceptionWorkingHours;
import models.facility.Accessibility;
import models.facility.ExamMachine;
import models.facility.ExamRoom;
import models.facility.ExamStartingHour;
import models.facility.MailAddress;
import models.user.Role;
import models.user.User;
import org.apache.pekko.actor.ActorSystem;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.data.Form;
import play.libs.Json;
import play.mvc.Http;
import play.mvc.Result;
import sanitizers.Attrs;
import scala.concurrent.duration.Duration;
import security.Authenticated;
import system.interceptors.SensitiveDataPolicy;

public class RoomController extends BaseController {

    private final boolean examVisitActivated;
    private final String defaultTimeZoneId;
    private final ExternalFacilityAPI externalApi;
    protected ActorSystem system;
    private final DateTimeHandler dateTimeHandler;
    private final Logger logger = LoggerFactory.getLogger(RoomController.class);
    private static final int EPOCH = 1970;

    @Inject
    public RoomController(
        ExternalFacilityAPI externalFacilityAPI,
        ActorSystem actorSystem,
        ConfigReader configReader,
        DateTimeHandler dateTimeHandler
    ) {
        this.externalApi = externalFacilityAPI;
        this.system = actorSystem;
        this.examVisitActivated = configReader.isVisitingExaminationSupported();
        this.defaultTimeZoneId = configReader.getDefaultTimeZone().getID();
        this.dateTimeHandler = dateTimeHandler;
    }

    private CompletionStage<Result> updateRemote(ExamRoom room) throws MalformedURLException {
        if (room.getExternalRef() != null && examVisitActivated) {
            return externalApi
                .updateFacility(room)
                .thenApplyAsync(x -> ok(Json.newObject()))
                .exceptionally(throwable -> internalServerError(Json.newObject().put("error", throwable.getMessage())));
        } else {
            return wrapAsPromise(ok());
        }
    }

    private void asyncUpdateRemote(ExamRoom room) {
        // Handle remote updates in dedicated threads
        if (room.getExternalRef() != null && examVisitActivated) {
            system
                .scheduler()
                .scheduleOnce(
                    Duration.create(1, TimeUnit.SECONDS),
                    () -> {
                        try {
                            externalApi.updateFacility(room);
                        } catch (MalformedURLException e) {
                            logger.error("Remote update of exam room #{} failed", room.getExternalRef());
                        }
                    },
                    system.dispatcher()
                );
        }
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("SUPPORT"), @Group("ADMIN"), @Group("STUDENT") })
    @SensitiveDataPolicy(sensitiveFieldNames = { "internalPassword", "externalPassword" })
    public Result getExamRooms(Http.Request request) {
        ExpressionList<ExamRoom> query = DB.find(ExamRoom.class)
            .fetch("accessibilities")
            .fetch("examMachines")
            .fetch("defaultWorkingHours")
            .fetch("calendarExceptionEvents")
            .fetch("examStartingHours")
            .where();
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (!user.hasRole(Role.Name.ADMIN)) {
            query = query.ne("state", ExamRoom.State.INACTIVE.toString());
        }
        List<ExamRoom> rooms = query.findList();
        for (ExamRoom room : rooms) {
            room.getExamMachines().removeIf(ExamMachine::isArchived);
            room.setInternalPasswordRequired(room.getInternalPassword() != null);
            room.setExternalPasswordRequired(room.getExternalPassword() != null);
        }
        PathProperties props = PathProperties.parse(
            "(*, mailAddress(*), accessibilities(*), defaultWorkingHours(*), calendarExceptionEvents(*), examStartingHours(*), examMachines(*, softwareInfo(*)))"
        );
        return ok(rooms, props);
    }

    @Restrict(@Group("ADMIN"))
    public Result getExamRoom(Long id) {
        ExamRoom examRoom = DB.find(ExamRoom.class, id);
        if (examRoom == null) {
            return notFound("room not found");
        }
        PathProperties props = PathProperties.parse(
            "(*, defaultWorkingHours(*), calendarExceptionEvents(*), accessibilities(*), mailAddress(*), examStartingHours(*), examMachines(*))"
        );
        return ok(examRoom, props);
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result createExamRoomDraft() {
        ExamRoom examRoom = new ExamRoom();
        examRoom.setState("SAVED");
        examRoom.setLocalTimezone(defaultTimeZoneId);
        examRoom.setMailAddress(new MailAddress());
        examRoom.save();
        return ok(Json.toJson(examRoom));
    }

    @Restrict({ @Group("ADMIN"), @Group("STUDENT") })
    public Result validatePassword(Long roomId, Http.Request request) {
        ExamRoom room = DB.find(ExamRoom.class, roomId);
        if (room == null) {
            return notFound("room not found");
        }

        if (room.getInternalPassword() == null) {
            return badRequest("room does not require internal password");
        }
        JsonNode body = request.body().asJson();
        boolean isExternal = body.get("external").asBoolean();
        String validPassword = isExternal ? room.getExternalPassword() : room.getInternalPassword();
        if (validPassword.equals(body.get("password").asText())) {
            return ok();
        } else {
            return forbidden();
        }
    }

    @Restrict(@Group({ "ADMIN" }))
    public CompletionStage<Result> updateExamRoom(Long id, Http.Request request) throws MalformedURLException {
        ExamRoom room = formFactory
            .form(ExamRoom.class)
            .bindFromRequest(
                request,
                "name",
                "roomCode",
                "buildingName",
                "campus",
                "accessibilityInfo",
                "accessible",
                "roomInstruction",
                "roomInstructionEN",
                "roomInstructionSV",
                "contactPerson",
                "videoRecordingsURL",
                "examMachineCount",
                "statusComment",
                "outOfService",
                "state",
                "internalPassword",
                "externalPassword"
            )
            .get();
        ExamRoom existing = DB.find(ExamRoom.class, id);
        if (existing == null) {
            return wrapAsPromise(notFound());
        }
        existing.setName(room.getName());
        existing.setRoomCode(room.getRoomCode());
        existing.setBuildingName(room.getBuildingName());
        existing.setCampus(room.getCampus());
        existing.setAccessible(room.getAccessible());
        existing.setRoomInstruction(room.getRoomInstruction());
        existing.setRoomInstructionEN(room.getRoomInstructionEN());
        existing.setRoomInstructionSV(room.getRoomInstructionSV());
        existing.setContactPerson(room.getContactPerson());
        existing.setVideoRecordingsURL(room.getVideoRecordingsURL());
        existing.setStatusComment(room.getStatusComment());
        existing.setOutOfService(room.getOutOfService());
        existing.setState(room.getState());
        existing.setInternalPassword(room.getInternalPassword());
        existing.setExternalPassword(room.getExternalPassword());

        existing.update();

        return updateRemote(existing);
    }

    @Restrict(@Group({ "ADMIN" }))
    public CompletionStage<Result> updateExamRoomAddress(Long id, Http.Request request) throws MalformedURLException {
        MailAddress address = bindAddress(request);
        MailAddress existing = DB.find(MailAddress.class, id);
        if (existing == null) {
            return wrapAsPromise(notFound());
        }
        Optional<ExamRoom> room = DB.find(ExamRoom.class).where().eq("mailAddress", existing).findOneOrEmpty();
        if (room.isEmpty()) {
            return wrapAsPromise(notFound());
        }
        existing.setCity(address.getCity());
        existing.setStreet(address.getStreet());
        existing.setZip(address.getZip());
        existing.update();
        return updateRemote(room.get());
    }

    private MailAddress bindAddress(Http.Request request) {
        final Form<MailAddress> form = formFactory.form(MailAddress.class);
        return form.bindFromRequest(request).get();
    }

    private DefaultWorkingHours parseWorkingHours(JsonNode root) {
        JsonNode node = root.get("workingHours");
        DateTimeFormatter formatter = ISODateTimeFormat.dateTimeParser();
        DefaultWorkingHours dwh = new DefaultWorkingHours();
        dwh.setWeekday(node.get("weekday").asText());
        // Deliberately use Jan to have no DST in effect
        DateTime startTime = DateTime.parse(node.get("startTime").asText(), formatter).withDayOfYear(1).withYear(EPOCH);
        DateTime endTime = DateTime.parse(node.get("endTime").asText(), formatter).withDayOfYear(1).withYear(EPOCH);
        dwh.setStartTime(startTime);
        dwh.setEndTime(endTime);
        return dwh;
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result updateExamRoomWorkingHours(Http.Request request) {
        JsonNode root = request.body().asJson();
        List<Long> roomIds = new ArrayList<>();
        for (JsonNode roomId : root.get("roomIds")) {
            roomIds.add(roomId.asLong());
        }
        List<ExamRoom> rooms = DB.find(ExamRoom.class).where().idIn(roomIds).findList();
        DefaultWorkingHours hours = parseWorkingHours(root);
        for (ExamRoom examRoom : rooms) {
            // Find out if there's overlap. Remove those
            List<DefaultWorkingHours> existing = DB.find(DefaultWorkingHours.class)
                .where()
                .eq("room", examRoom)
                .eq("weekday", hours.getWeekday())
                .findList();
            List<DefaultWorkingHours> overlapping = existing
                .stream()
                .filter(dwh -> dwh.overlaps(hours))
                .toList();
            DB.deleteAll(overlapping);
            examRoom.getDefaultWorkingHours().removeAll(overlapping);

            hours.setRoom(examRoom);
            DateTime end = new DateTime(hours.getEndTime());
            int offset = DateTimeZone.forID(examRoom.getLocalTimezone()).getOffset(end);
            int endMillisOfDay = dateTimeHandler.resolveEndWorkingHourMillis(end, offset) - offset;
            hours.setEndTime(end.withMillisOfDay(endMillisOfDay));
            hours.setTimezoneOffset(offset);
            hours.save();
            examRoom.getDefaultWorkingHours().add(hours);
            asyncUpdateRemote(examRoom);
        }
        return ok(Json.newObject().put("id", hours.getId()));
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result removeExamRoomWorkingHours(Long roomId, Long id) {
        DefaultWorkingHours dwh = DB.find(DefaultWorkingHours.class, id);
        ExamRoom room = DB.find(ExamRoom.class, roomId);
        if (dwh == null || room == null) {
            return forbidden();
        }
        dwh.delete();
        room.getDefaultWorkingHours().remove(dwh);
        asyncUpdateRemote(room);
        return ok();
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result updateExamStartingHours(Http.Request request) {
        JsonNode root = request.body().asJson();
        List<Long> roomIds = new ArrayList<>();
        for (JsonNode roomId : root.get("roomIds")) {
            roomIds.add(roomId.asLong());
        }

        List<ExamRoom> rooms = DB.find(ExamRoom.class).where().idIn(roomIds).findList();

        for (ExamRoom examRoom : rooms) {
            if (examRoom == null) {
                return notFound();
            }
            List<ExamStartingHour> previous = DB.find(ExamStartingHour.class)
                .where()
                .eq("room.id", examRoom.getId())
                .findList();
            DB.deleteAll(previous);

            JsonNode node = request.body().asJson();
            DateTimeFormatter formatter = DateTimeFormat.forPattern("dd.MM.yyyy HH:mmZZ");
            for (JsonNode hours : node.get("hours")) {
                ExamStartingHour esh = new ExamStartingHour();
                esh.setRoom(examRoom);
                // Deliberately use first/second of Jan to have no DST in effect
                DateTime startTime = DateTime.parse(hours.asText(), formatter).withDayOfYear(1);
                esh.setStartingHour(startTime.toDate());
                esh.setTimezoneOffset(DateTimeZone.forID(examRoom.getLocalTimezone()).getOffset(startTime));

                esh.save();
            }
            asyncUpdateRemote(examRoom);
        }
        return ok();
    }

    private ExceptionWorkingHours parse(JsonNode node) {
        DateTime startDate = ISODateTimeFormat.dateTime().parseDateTime(node.get("start").asText());
        DateTime endDate = ISODateTimeFormat.dateTime().parseDateTime(node.get("end").asText());
        ExceptionWorkingHours hours = new ExceptionWorkingHours();
        hours.setStartDate(startDate.toDate());
        hours.setEndDate(endDate.toDate());
        hours.setOutOfService(node.get("outOfService").asBoolean(true));
        return hours;
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result addRoomExceptionHours(Http.Request request) {
        final JsonNode exceptionsNode = request.body().asJson().get("exceptions");
        final JsonNode ids = request.body().asJson().get("roomIds");

        List<Long> roomIds = StreamSupport.stream(ids.spliterator(), false).map(JsonNode::asLong).toList();
        List<ExamRoom> rooms = DB.find(ExamRoom.class).where().idIn(roomIds).findList();
        for (ExamRoom room : rooms) {
            for (JsonNode node : exceptionsNode) {
                ExceptionWorkingHours exception = parse(node);
                exception.setStartDateTimezoneOffset(
                    DateTimeZone.forID(room.getLocalTimezone()).getOffset(exception.getStartDate().getTime())
                );
                exception.setEndDateTimezoneOffset(
                    DateTimeZone.forID(room.getLocalTimezone()).getOffset(exception.getEndDate().getTime())
                );
                exception.setRoom(room);
                exception.save();
                room.getCalendarExceptionEvents().add(exception);
            }
            asyncUpdateRemote(room);
        }
        return ok(
            Json.toJson(
                rooms
                    .stream()
                    .flatMap(r -> r.getCalendarExceptionEvents().stream())
                    .toList()
            )
        );
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result updateExamRoomAccessibility(Long id, Http.Request request) {
        JsonNode json = request.body().asJson();
        final List<String> ids = Arrays.asList(json.get("ids").asText().split(","));

        ExamRoom room = DB.find(ExamRoom.class, id);
        if (room == null) {
            return notFound();
        }
        room.getAccessibilities().clear();
        room.save();

        if (!ids.isEmpty()) {
            for (String aid : ids) {
                int accessibilityId;
                try {
                    accessibilityId = Integer.parseInt(aid.trim());
                } catch (Exception ex) {
                    break;
                }
                Accessibility accessibility = DB.find(Accessibility.class, accessibilityId);
                room.getAccessibilities().add(accessibility);
                room.save();
                asyncUpdateRemote(room);
            }
        }
        return ok();
    }

    @Restrict(@Group({ "ADMIN" }))
    public CompletionStage<Result> removeRoomExceptionHour(Long roomId, Long exceptionId) throws MalformedURLException {
        ExceptionWorkingHours exception = DB.find(ExceptionWorkingHours.class, exceptionId);
        ExamRoom room = DB.find(ExamRoom.class, roomId);
        if (exception == null || room == null) {
            return wrapAsPromise(notFound());
        }
        exception.delete();
        room.getCalendarExceptionEvents().remove(exception);
        return updateRemote(room);
    }

    @Restrict(@Group({ "ADMIN" }))
    public CompletionStage<Result> inactivateExamRoom(Long id) throws MalformedURLException {
        ExamRoom room = DB.find(ExamRoom.class, id);
        if (room == null) {
            return wrapAsPromise(notFound());
        }
        room.setState(ExamRoom.State.INACTIVE.toString());
        room.update();

        if (room.getExternalRef() != null && examVisitActivated) {
            return externalApi.inactivateFacility(room.getId()).thenApplyAsync(response -> ok(Json.toJson(room)));
        } else {
            return wrapAsPromise(ok(Json.toJson(room)));
        }
    }

    @Restrict(@Group({ "ADMIN" }))
    public CompletionStage<Result> activateExamRoom(Long id) throws MalformedURLException {
        ExamRoom room = DB.find(ExamRoom.class, id);
        if (room == null) {
            return wrapAsPromise(notFound());
        }
        room.setState(ExamRoom.State.ACTIVE.toString());
        room.update();

        if (room.getExternalRef() != null && examVisitActivated) {
            return externalApi.activateFacility(room.getId()).thenApplyAsync(response -> ok(Json.toJson(room)));
        } else {
            return wrapAsPromise(ok(Json.toJson(room)));
        }
    }
}

File: app/controllers/facility/RoomController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.facility

import impl.FacilityHandler
import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.cache.FacilityCache
import miscellaneous.config.ConfigReader
import miscellaneous.datetime.DateTimeHandler
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.calendar.{DefaultWorkingHours, ExceptionWorkingHours}
import models.facility.{Accessibility, ExamRoom, ExamStartingHour, MailAddress}
import models.user.Role
import org.apache.pekko.actor.ActorSystem
import org.joda.time.format.{DateTimeFormat, ISODateTimeFormat}
import org.joda.time.{DateTime, DateTimeZone}
import play.api.Logging
import play.api.libs.json.{JsArray, JsValue, Json}
import play.api.mvc.*
import security.scala.Auth
import security.scala.Auth.{AuthenticatedAction, authorized}
import system.AuditedAction
import system.interceptors.scala.SensitiveDataFilter

import javax.inject.Inject
import scala.concurrent.duration.*
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters.*

class RoomController @Inject() (
    facilityHandler: FacilityHandler,
    actorSystem: ActorSystem,
    configReader: ConfigReader,
    dateTimeHandler: DateTimeHandler,
    facilityCache: FacilityCache,
    authenticated: AuthenticatedAction,
    audited: AuditedAction,
    sensitiveDataFilter: SensitiveDataFilter,
    val controllerComponents: ControllerComponents
)(implicit ec: ExecutionContext)
    extends BaseController
    with DbApiHelper
    with JavaApiHelper
    with Logging:

  private val examVisitActivated = configReader.isVisitingExaminationSupported
  private val defaultTimeZoneId  = configReader.getDefaultTimeZone.getID
  private val EPOCH              = 1970

  private def updateRemote(room: ExamRoom): Future[Result] =
    if Option(room.getExternalRef).isDefined && examVisitActivated then
      facilityHandler
        .updateFacility(room)
        .map(_ => Ok(Json.obj()))
        .recover { case ex: Throwable =>
          InternalServerError(Json.obj("error" -> ex.getMessage))
        }
    else Future.successful(Ok)

  private def asyncUpdateRemote(room: ExamRoom): Unit =
    // Handle remote updates in dedicated threads
    if Option(room.getExternalRef).isDefined && examVisitActivated then
      actorSystem.scheduler.scheduleOnce(1.second) {
        facilityHandler.updateFacility(room).recover { case e: Exception =>
          logger.error(s"Remote update of exam room #${room.getExternalRef} failed", e)
        }
      }

  def getExamRooms: Action[AnyContent] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.SUPPORT, Role.Name.ADMIN, Role.Name.STUDENT)))
      .andThen(sensitiveDataFilter(Set("internalPassword", "externalPassword"))) { request =>
        val user = request.attrs(Auth.ATTR_USER)
        var query = DB
          .find(classOf[ExamRoom])
          .fetch("accessibilities")
          .fetch("examMachines")
          .fetch("defaultWorkingHours")
          .fetch("calendarExceptionEvents")
          .fetch("examStartingHours")
          .where()

        if !user.hasRole(Role.Name.ADMIN) then query = query.ne("state", ExamRoom.State.INACTIVE.toString)

        val rooms = query.list
        rooms.foreach { room =>
          room.getExamMachines.removeIf(_.isArchived)
          room.setInternalPasswordRequired(Option(room.getInternalPassword).isDefined)
          room.setExternalPasswordRequired(Option(room.getExternalPassword).isDefined)
        }

        val props = PathProperties.parse(
          "(*, mailAddress(*), accessibilities(*), defaultWorkingHours(*), calendarExceptionEvents(*), examStartingHours(*), examMachines(*, softwareInfo(*)))"
        )
        Ok(rooms.asJson(props))
      }

  def getExamRoom(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      Option(DB.find(classOf[ExamRoom], id)) match
        case None => NotFound("room not found")
        case Some(examRoom) =>
          val props = PathProperties.parse(
            "(*, defaultWorkingHours(*), calendarExceptionEvents(*), accessibilities(*), mailAddress(*), examStartingHours(*), examMachines(*))"
          )
          Ok(examRoom.asJson(props))
    }

  def createExamRoomDraft: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val examRoom = new ExamRoom()
      examRoom.setState("SAVED")
      examRoom.setLocalTimezone(defaultTimeZoneId)
      examRoom.setMailAddress(new MailAddress())
      examRoom.save()
      Ok(examRoom.asJson)
    }

  def validatePassword(roomId: Long): Action[JsValue] = audited
    .andThen(authenticated)(parse.json)
    .andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.STUDENT))) { request =>
      val body = request.body

      // Check if this is an external facility validation
      val isExternalFacility = (body \ "external").asOpt[Boolean].getOrElse(false)
      if isExternalFacility then validateExternalFacilityPassword(body)
      else validateInternalRoomPassword(roomId, body)
    }

  private def validateExternalFacilityPassword(body: JsValue): Result =
    (body \ "facilityId").asOpt[String] match
      case None => BadRequest("facilityId is required")
      case Some(facilityId) =>
        facilityCache.getFacilityPassword(facilityId) match
          case None => NotFound("Facility not found")
          case Some(facilityPassword) =>
            (body \ "password").asOpt[String] match
              case Some(providedPassword) if facilityPassword == providedPassword => Ok
              case _ => Forbidden("Invalid password for facility")

  private def validateInternalRoomPassword(roomId: Long, body: JsValue): Result =
    Option(DB.find(classOf[ExamRoom], roomId)) match
      case None => NotFound("room not found")
      case Some(room) =>
        (body \ "password").asOpt[String] match
          case Some(password) if Option(room.getInternalPassword).contains(password) => Ok
          case _                                                                     => Forbidden("Invalid password")

  def updateExamRoom(id: Long): Action[JsValue] = audited
    .andThen(authenticated)(parse.json)
    .andThen(authorized(Seq(Role.Name.ADMIN)))
    .async { request =>
      Option(DB.find(classOf[ExamRoom], id)) match
        case None => Future.successful(NotFound("room not found"))
        case Some(existing) =>
          val body = request.body
          (body \ "name").asOpt[String].foreach(existing.setName)
          (body \ "roomCode").asOpt[String].foreach(existing.setRoomCode)
          (body \ "buildingName").asOpt[String].foreach(existing.setBuildingName)
          (body \ "campus").asOpt[String].foreach(existing.setCampus)
          (body \ "accessible").asOpt[Boolean].foreach(v => existing.setAccessible(v))
          (body \ "roomInstruction").asOpt[String].foreach(existing.setRoomInstruction)
          (body \ "roomInstructionEN").asOpt[String].foreach(existing.setRoomInstructionEN)
          (body \ "roomInstructionSV").asOpt[String].foreach(existing.setRoomInstructionSV)
          (body \ "contactPerson").asOpt[String].foreach(existing.setContactPerson)
          (body \ "videoRecordingsURL").asOpt[String].foreach(existing.setVideoRecordingsURL)
          (body \ "statusComment").asOpt[String].foreach(existing.setStatusComment)
          (body \ "outOfService").asOpt[Boolean].foreach(v => existing.setOutOfService(v))
          (body \ "state").asOpt[String].foreach(existing.setState)
          (body \ "internalPassword").asOpt[String].foreach(existing.setInternalPassword)
          (body \ "externalPassword").asOpt[String].foreach(existing.setExternalPassword)

          existing.update()
          updateRemote(existing)
    }

  def updateExamRoomAddress(id: Long): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.ADMIN)))(parse.json)
    .async { request =>
      Option(DB.find(classOf[MailAddress], id)) match
        case None => Future.successful(NotFound("address not found"))
        case Some(existing) =>
          DB.find(classOf[ExamRoom]).where().eq("mailAddress", existing).find match
            case None => Future.successful(NotFound("room not found"))
            case Some(room) =>
              val body = request.body
              (body \ "city").asOpt[String].foreach(existing.setCity)
              (body \ "street").asOpt[String].foreach(existing.setStreet)
              (body \ "zip").asOpt[String].foreach(existing.setZip)
              existing.update()
              updateRemote(room)
    }

  private def parseWorkingHours(body: JsValue): DefaultWorkingHours =
    val node      = body \ "workingHours"
    val formatter = ISODateTimeFormat.dateTimeParser()
    val dwh       = new DefaultWorkingHours()
    dwh.setWeekday((node \ "weekday").as[String])
    // Deliberately use Jan to have no DST in effect
    val startTime = DateTime.parse((node \ "startTime").as[String], formatter).withDayOfYear(1).withYear(EPOCH)
    val endTime   = DateTime.parse((node \ "endTime").as[String], formatter).withDayOfYear(1).withYear(EPOCH)
    dwh.setStartTime(startTime)
    dwh.setEndTime(endTime)
    dwh

  def updateExamRoomWorkingHours(): Action[JsValue] =
    authenticated(parse.json).andThen(authorized(Seq(Role.Name.ADMIN))) { request =>
      val body    = request.body
      val roomIds = (body \ "roomIds").as[List[Long]]
      val rooms   = DB.find(classOf[ExamRoom]).where().idIn(roomIds.map(Long.box).asJava).list
      val hours   = parseWorkingHours(body)

      rooms.foreach { examRoom =>
        // Find out if there's any overlap. Remove those
        val existing = DB
          .find(classOf[DefaultWorkingHours])
          .where()
          .eq("room", examRoom)
          .eq("weekday", hours.getWeekday)
          .list

        val overlapping = existing.filter(_.overlaps(hours))
        DB.deleteAll(overlapping.asJava)
        overlapping.foreach(examRoom.getDefaultWorkingHours.remove)

        hours.setRoom(examRoom)
        val end            = new DateTime(hours.getEndTime)
        val offset         = DateTimeZone.forID(examRoom.getLocalTimezone).getOffset(end)
        val endMillisOfDay = dateTimeHandler.resolveEndWorkingHourMillis(end, offset) - offset
        hours.setEndTime(end.withMillisOfDay(endMillisOfDay))
        hours.setTimezoneOffset(offset)
        hours.save()
        examRoom.getDefaultWorkingHours.add(hours)
        asyncUpdateRemote(examRoom)
      }

      Ok(Json.obj("id" -> hours.getId.longValue))
    }

  def removeExamRoomWorkingHours(roomId: Long, id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      (Option(DB.find(classOf[DefaultWorkingHours], id)), Option(DB.find(classOf[ExamRoom], roomId))) match
        case (Some(dwh), Some(room)) =>
          dwh.delete()
          room.getDefaultWorkingHours.remove(dwh)
          asyncUpdateRemote(room)
          Ok
        case _ => Forbidden("working hours or room not found")
    }

  def updateExamStartingHours(): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.ADMIN)))(parse.json) { request =>
      val body      = request.body
      val roomIds   = (body \ "roomIds").as[List[Long]]
      val rooms     = DB.find(classOf[ExamRoom]).where().idIn(roomIds.map(Long.box).asJava).list
      val formatter = DateTimeFormat.forPattern("dd.MM.yyyy HH:mmZZ")

      rooms.foreach { examRoom =>
        val previous = DB.find(classOf[ExamStartingHour]).where().eq("room.id", examRoom.getId).list
        DB.deleteAll(previous.asJava)

        (body \ "hours").as[JsArray].value.foreach { hoursNode =>
          val esh = new ExamStartingHour()
          esh.setRoom(examRoom)
          // Deliberately use the first/second of Jan to have no DST in effect
          val startTime = DateTime.parse(hoursNode.as[String], formatter).withDayOfYear(1)
          esh.setStartingHour(startTime.toDate)
          esh.setTimezoneOffset(DateTimeZone.forID(examRoom.getLocalTimezone).getOffset(startTime))
          esh.save()
        }
        asyncUpdateRemote(examRoom)
      }

      Ok
    }

  private def parseException(node: JsValue): ExceptionWorkingHours =
    val startDate = ISODateTimeFormat.dateTime().parseDateTime((node \ "startDate").as[String])
    val endDate   = ISODateTimeFormat.dateTime().parseDateTime((node \ "endDate").as[String])
    val hours     = new ExceptionWorkingHours()
    hours.setStartDate(startDate.toDate)
    hours.setEndDate(endDate.toDate)
    hours.setOutOfService((node \ "outOfService").asOpt[Boolean].getOrElse(true))
    hours

  def addRoomExceptionHours(): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.ADMIN)))(parse.json) { request =>
      val body           = request.body
      val exceptionsNode = (body \ "exceptions").as[JsArray]
      val roomIds        = (body \ "roomIds").as[List[Long]]
      val rooms          = DB.find(classOf[ExamRoom]).where().idIn(roomIds.map(Long.box).asJava).list

      rooms.foreach { room =>
        exceptionsNode.value.foreach { node =>
          val exception = parseException(node)
          exception.setStartDateTimezoneOffset(
            DateTimeZone.forID(room.getLocalTimezone).getOffset(exception.getStartDate.getTime)
          )
          exception.setEndDateTimezoneOffset(
            DateTimeZone.forID(room.getLocalTimezone).getOffset(exception.getEndDate.getTime)
          )
          exception.setRoom(room)
          exception.save()
          room.getCalendarExceptionEvents.add(exception)
        }
        asyncUpdateRemote(room)
      }

      val allExceptions = rooms.flatMap(_.getCalendarExceptionEvents.asScala)
      Ok(allExceptions.asJson)
    }

  def updateExamRoomAccessibility(id: Long): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.ADMIN)))(parse.json) { request =>
      Option(DB.find(classOf[ExamRoom], id)) match
        case None => NotFound("room not found")
        case Some(room) =>
          val ids = (request.body \ "ids").as[String].split(",").toList.map(_.trim)
          room.getAccessibilities.clear()
          room.save()

          if ids.nonEmpty && ids.head.nonEmpty then
            ids.foreach { aid =>
              scala.util.Try(aid.toInt).foreach { accessibilityId =>
                Option(DB.find(classOf[Accessibility], accessibilityId)).foreach { accessibility =>
                  room.getAccessibilities.add(accessibility)
                  room.save()
                  asyncUpdateRemote(room)
                }
              }
            }
          Ok
    }

  def removeRoomExceptionHour(roomId: Long, exceptionId: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))).async { _ =>
      (Option(DB.find(classOf[ExceptionWorkingHours], exceptionId)), Option(DB.find(classOf[ExamRoom], roomId))) match
        case (Some(exception), Some(room)) =>
          exception.delete()
          room.getCalendarExceptionEvents.remove(exception)
          updateRemote(room)
        case _ => Future.successful(NotFound("exception or room not found"))
    }

  def inactivateExamRoom(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))).async { _ =>
      Option(DB.find(classOf[ExamRoom], id)) match
        case None => Future.successful(NotFound("room not found"))
        case Some(room) =>
          room.setState(ExamRoom.State.INACTIVE.toString)
          room.update()

          if Option(room.getExternalRef).isDefined && examVisitActivated then
            facilityHandler.inactivateFacility(room.getId).map(_ => Ok(room.asJson))
          else Future.successful(Ok(room.asJson))
    }

  def activateExamRoom(id: Long): Action[AnyContent] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.ADMIN)))
    .async { _ =>
      Option(DB.find(classOf[ExamRoom], id)) match
        case None => Future.successful(NotFound("room not found"))
        case Some(room) =>
          room.setState(ExamRoom.State.ACTIVE.toString)
          room.update()

          if Option(room.getExternalRef).isDefined && examVisitActivated then
            facilityHandler.activateFacility(room.getId).map(_ => Ok(room.asJson))
          else Future.successful(Ok(room.asJson))
    }
